File: src/cccc/ports/mcp/utils/help_markdown.py
from __future__ import annotations

import re
from typing import Any, Optional


_HELP_ROLE_HEADER_RE = re.compile(r"^##\s*@role:\s*(\w+)\s*$", re.IGNORECASE)
_HELP_ACTOR_HEADER_RE = re.compile(r"^##\s*@actor:\s*(\S+)(?:\s+(.*\S))?\s*$", re.IGNORECASE)
_HELP_PET_HEADER_RE = re.compile(r"^##\s*@pet\s*:?\s*$", re.IGNORECASE)
_HELP_H2_RE = re.compile(r"^##(?!#)\s+.*$")
_HELP_LEGACY_ROLE_SECTION_RE = re.compile(r"^##\s+Role Notes\s*$", re.IGNORECASE)
_HELP_H3_RE = re.compile(r"^###\s+(.+?)\s*$")


def _split_sections(markdown: str) -> list[str]:
    raw = str(markdown or "").replace("\r\n", "\n").replace("\r", "\n")
    if not raw:
        return [""]
    lines = raw.split("\n")
    sections: list[str] = []
    current: list[str] = []
    for line in lines:
        if _HELP_H2_RE.match(line) and current:
            sections.append("\n".join(current))
            current = [line]
            continue
        current.append(line)
    sections.append("\n".join(current))
    return sections


def _trim_block(text: str) -> str:
    return str(text or "").strip()


def _parse_tagged_section(section: str) -> Optional[dict[str, str]]:
    normalized = str(section or "").replace("\r\n", "\n").replace("\r", "\n")
    lines = normalized.split("\n")
    header = str(lines[0] or "")
    role_match = _HELP_ROLE_HEADER_RE.match(header)
    if role_match:
        role = str(role_match.group(1) or "").strip().lower()
        body = _trim_block("\n".join(lines[1:]))
        if role in {"foreman", "peer"}:
            return {"kind": "role", "key": f"role:{role}", "raw": _trim_block(normalized), "body": body}
        return {"kind": "extra", "key": f"role:{role}", "raw": _trim_block(normalized), "body": body}
    actor_match = _HELP_ACTOR_HEADER_RE.match(header)
    if actor_match:
        actor_id = str(actor_match.group(1) or "").strip()
        inline_body = str(actor_match.group(2) or "").strip()
        body_lines = [inline_body] if inline_body else []
        body_lines.extend(lines[1:])
        return {
            "kind": "actor" if actor_id else "extra",
            "key": f"actor:{actor_id}" if actor_id else "actor:",
            "raw": _trim_block(normalized),
            "body": _trim_block("\n".join(body_lines)),
        }
    if _HELP_PET_HEADER_RE.match(header):
        return {
            "kind": "pet",
            "key": "pet",
            "raw": _trim_block(normalized),
            "body": _trim_block("\n".join(lines[1:])),
        }
    return None


def _try_extract_legacy_role_notes(common: str) -> dict[str, Any]:
    normalized = str(common or "").replace("\r\n", "\n").replace("\r", "\n")
    sections = _split_sections(normalized)
    kept: list[str] = []
    foreman = ""
    peer = ""
    used = False

    for section in sections:
        raw = _trim_block(section)
        if not raw:
            continue
        lines = raw.split("\n")
        if not _HELP_LEGACY_ROLE_SECTION_RE.match(str(lines[0] or "")):
            kept.append(raw)
            continue
        body_lines = lines[1:]
        chunks: list[dict[str, str]] = []
        current_title = ""
        current_body: list[str] = []
        stray = False
        for line in body_lines:
            h3 = _HELP_H3_RE.match(line)
            if h3:
                if current_title:
                    chunks.append({"title": current_title, "body": _trim_block("\n".join(current_body))})
                elif _trim_block("\n".join(current_body)):
                    stray = True
                current_title = str(h3.group(1) or "").strip()
                current_body = []
                continue
            current_body.append(line)
        if current_title:
            chunks.append({"title": current_title, "body": _trim_block("\n".join(current_body))})
        elif _trim_block("\n".join(current_body)):
            stray = True
        if stray or not chunks:
            kept.append(raw)
            continue
        local_foreman = ""
        local_peer = ""
        unknown = False
        for chunk in chunks:
            title = str(chunk.get("title") or "").strip().lower()
            if title == "foreman":
                local_foreman = str(chunk.get("body") or "")
            elif title == "peer":
                local_peer = str(chunk.get("body") or "")
            else:
                unknown = True
                break
        if unknown:
            kept.append(raw)
            continue
        foreman = local_foreman
        peer = local_peer
        used = True

    return {
        "common": "\n\n".join(kept).strip(),
        "foreman": foreman,
        "peer": peer,
        "used_legacy_role_notes": used,
    }


def parse_help_markdown(markdown: str) -> dict[str, Any]:
    sections = _split_sections(markdown)
    common_sections: list[str] = []
    actor_notes: dict[str, str] = {}
    extra_tagged_blocks: list[str] = []
    foreman = ""
    peer = ""
    pet = ""

    for section in sections:
        raw = _trim_block(section)
        if not raw:
            continue
        tagged = _parse_tagged_section(raw)
        if not tagged:
            common_sections.append(raw)
            continue
        kind = str(tagged.get("kind") or "")
        key = str(tagged.get("key") or "")
        body = str(tagged.get("body") or "")
        if kind == "role":
            if key == "role:foreman":
                foreman = body
            elif key == "role:peer":
                peer = body
            else:
                extra_tagged_blocks.append(str(tagged.get("raw") or ""))
            continue
        if kind == "actor":
            actor_id = key[len("actor:") :]
            if actor_id:
                actor_notes[actor_id] = body
            else:
                extra_tagged_blocks.append(str(tagged.get("raw") or ""))
            continue
        if kind == "pet":
            pet = body
            continue
        extra_tagged_blocks.append(str(tagged.get("raw") or ""))

    common = "\n\n".join(common_sections).strip()
    used_legacy_role_notes = False
    if not foreman and not peer:
        legacy = _try_extract_legacy_role_notes(common)
        common = str(legacy.get("common") or "")
        foreman = str(legacy.get("foreman") or "")
        peer = str(legacy.get("peer") or "")
        used_legacy_role_notes = bool(legacy.get("used_legacy_role_notes"))

    return {
        "common": common,
        "foreman": foreman,
        "peer": peer,
        "pet": pet,
        "actor_notes": actor_notes,
        "extra_tagged_blocks": extra_tagged_blocks,
        "used_legacy_role_notes": used_legacy_role_notes,
    }


def build_help_markdown(
    *,
    common: str,
    foreman: str,
    peer: str,
    pet: str,
    actor_notes: dict[str, str],
    actor_order: Optional[list[str]] = None,
    extra_tagged_blocks: Optional[list[str]] = None,
) -> str:
    parts: list[str] = []
    common_text = _trim_block(common)
    foreman_text = _trim_block(foreman)
    peer_text = _trim_block(peer)
    pet_text = _trim_block(pet)
    actor_notes_map = dict(actor_notes or {})
    extra_blocks = [_trim_block(item) for item in list(extra_tagged_blocks or []) if _trim_block(item)]

    if common_text:
        parts.append(common_text)
    if foreman_text:
        parts.append(f"## @role: foreman\n\n{foreman_text}")
    if peer_text:
        parts.append(f"## @role: peer\n\n{peer_text}")
    if pet_text:
        parts.append(f"## @pet\n\n{pet_text}")

    seen: set[str] = set()
    ordered_actor_ids: list[str] = []
    for actor_id in list(actor_order or []):
        aid = str(actor_id or "").strip()
        if not aid or aid in seen:
            continue
        seen.add(aid)
        ordered_actor_ids.append(aid)
    for actor_id in sorted(actor_notes_map.keys()):
        if actor_id in seen:
            continue
        seen.add(actor_id)
        ordered_actor_ids.append(actor_id)
    for actor_id in ordered_actor_ids:
        body = _trim_block(str(actor_notes_map.get(actor_id) or ""))
        if not body:
            continue
        parts.append(f"## @actor: {actor_id}\n\n{body}")

    parts.extend(extra_blocks)
    out = "\n\n".join([part for part in parts if part]).strip()
    return f"{out}\n" if out else ""


def update_actor_help_note(markdown: str, actor_id: str, note: str, actor_order: Optional[list[str]] = None) -> str:
    parsed = parse_help_markdown(markdown)
    next_actor_notes = dict(parsed.get("actor_notes") or {})
    aid = str(actor_id or "").strip()
    if aid:
        next_actor_notes[aid] = _trim_block(note)
    if aid and not next_actor_notes.get(aid):
        next_actor_notes.pop(aid, None)
    return build_help_markdown(
        common=str(parsed.get("common") or ""),
        foreman=str(parsed.get("foreman") or ""),
        peer=str(parsed.get("peer") or ""),
        pet=str(parsed.get("pet") or ""),
        actor_notes=next_actor_notes,
        actor_order=actor_order,
        extra_tagged_blocks=list(parsed.get("extra_tagged_blocks") or []),
    )


def _select_help_markdown(markdown: str, *, role: Optional[str], actor_id: Optional[str], include_pet: bool = False) -> str:
    """Filter CCCC_HELP markdown by optional conditional blocks.

    Supported markers (level-2 headings):
    - "## @role: foreman|peer"
    - "## @actor: <actor_id>"
    - "## @pet"

    Untagged content is always included. Tagged blocks are filtered only when the selector is known.

    A tagged block starts at its marker heading and ends at the next level-2 heading.
    Within tagged blocks, prefer "###" for subheadings (so "##" can remain a block boundary).
    """
    raw = str(markdown or "")
    if not raw.strip():
        return raw

    role_norm = str(role or "").strip().casefold()
    actor_norm = str(actor_id or "").strip()
    lines = raw.splitlines()
    keep_trailing_newline = raw.endswith("\n")

    out: list[str] = []
    buf: list[str] = []
    tag_kind: Optional[str] = None
    tag_value: str = ""

    def _include_block() -> bool:
        if tag_kind is None:
            return True
        if tag_kind == "role":
            if not role_norm:
                return True
            return role_norm == str(tag_value or "").strip().casefold()
        if tag_kind == "actor":
            if not actor_norm:
                return False
            return actor_norm == str(tag_value or "").strip()
        if tag_kind == "pet":
            return bool(include_pet)
        return True

    def _flush() -> None:
        nonlocal buf
        if buf and _include_block():
            out.extend(buf)
        buf = []

    for ln in lines:
        m_role = _HELP_ROLE_HEADER_RE.match(ln)
        m_actor = _HELP_ACTOR_HEADER_RE.match(ln)
        m_pet = _HELP_PET_HEADER_RE.match(ln)
        is_h2 = bool(_HELP_H2_RE.match(ln))

        if m_role or m_actor or m_pet:
            _flush()
            if m_role:
                tag_kind = "role"
                tag_value = str(m_role.group(1) or "").strip()
                role_label = tag_value.strip().casefold()
                if role_label == "foreman":
                    ln = "## Foreman"
                elif role_label == "peer":
                    ln = "## Peer"
                else:
                    ln = f"## Role: {tag_value}"
            else:
                if m_actor:
                    tag_kind = "actor"
                    tag_value = str(m_actor.group(1) or "").strip()
                    ln = "## Notes for you"
                else:
                    tag_kind = "pet"
                    tag_value = ""
                    ln = "## Pet Persona"
            buf.append(ln)
            continue

        if is_h2:
            _flush()
            tag_kind = None
            tag_value = ""

        buf.append(ln)
    _flush()

    out_text = "\n".join(out)
    if keep_trailing_newline:
        out_text += "\n"
    return out_text

File: web/src/utils/helpMarkdown.ts
export type HelpChangedBlock = "common" | "role:foreman" | "role:peer" | `actor:${string}`;

export type ParsedHelpMarkdown = {
  common: string;
  foreman: string;
  peer: string;
  actorNotes: Record<string, string>;
  extraTaggedBlocks: string[];
  usedLegacyRoleNotes: boolean;
};

type TaggedSection = {
  kind: "role" | "actor" | "extra";
  key: string;
  raw: string;
  body: string;
};

const H2_RE = /^##(?!#)\s+.*$/;
const ROLE_TAG_RE = /^##\s*@role:\s*(\w+)\s*$/i;
const ACTOR_TAG_RE = /^##\s*@actor:\s*(.+?)\s*$/i;
const LEGACY_ROLE_SECTION_RE = /^##\s+Role Notes\s*$/i;
const H3_RE = /^###\s+(.+?)\s*$/;

function splitSections(markdown: string): string[] {
  const raw = String(markdown || "").replace(/\r\n?/g, "\n");
  if (!raw) return [""];
  const lines = raw.split("\n");
  const sections: string[] = [];
  let current: string[] = [];
  for (const line of lines) {
    if (H2_RE.test(line) && current.length > 0) {
      sections.push(current.join("\n"));
      current = [line];
      continue;
    }
    current.push(line);
  }
  sections.push(current.join("\n"));
  return sections;
}

function trimBlock(text: string): string {
  return String(text || "").trim();
}

function parseTaggedSection(section: string): TaggedSection | null {
  const normalized = String(section || "").replace(/\r\n?/g, "\n");
  const lines = normalized.split("\n");
  const header = String(lines[0] || "");
  const roleMatch = header.match(ROLE_TAG_RE);
  if (roleMatch) {
    const role = String(roleMatch[1] || "").trim().toLowerCase();
    const body = trimBlock(lines.slice(1).join("\n"));
    if (role === "foreman" || role === "peer") {
      return { kind: "role", key: `role:${role}`, raw: trimBlock(normalized), body };
    }
    return { kind: "extra", key: `role:${role}`, raw: trimBlock(normalized), body };
  }
  const actorMatch = header.match(ACTOR_TAG_RE);
  if (actorMatch) {
    const actorId = String(actorMatch[1] || "").trim();
    return {
      kind: actorId ? "actor" : "extra",
      key: actorId ? `actor:${actorId}` : "actor:",
      raw: trimBlock(normalized),
      body: trimBlock(lines.slice(1).join("\n")),
    };
  }
  return null;
}

function tryExtractLegacyRoleNotes(common: string): {
  common: string;
  foreman: string;
  peer: string;
  used: boolean;
} {
  const normalized = String(common || "").replace(/\r\n?/g, "\n");
  const sections = splitSections(normalized);
  const kept: string[] = [];
  let foreman = "";
  let peer = "";
  let used = false;

  for (const section of sections) {
    const raw = trimBlock(section);
    if (!raw) continue;
    const lines = raw.split("\n");
    if (!LEGACY_ROLE_SECTION_RE.test(String(lines[0] || ""))) {
      kept.push(raw);
      continue;
    }
    const bodyLines = lines.slice(1);
    const chunks: Array<{ title: string; body: string }> = [];
    let currentTitle = "";
    let currentBody: string[] = [];
    let stray = false;
    for (const line of bodyLines) {
      const h3 = line.match(H3_RE);
      if (h3) {
        if (currentTitle) {
          chunks.push({ title: currentTitle, body: trimBlock(currentBody.join("\n")) });
        } else if (trimBlock(currentBody.join("\n"))) {
          stray = true;
        }
        currentTitle = String(h3[1] || "").trim();
        currentBody = [];
        continue;
      }
      currentBody.push(line);
    }
    if (currentTitle) {
      chunks.push({ title: currentTitle, body: trimBlock(currentBody.join("\n")) });
    } else if (trimBlock(currentBody.join("\n"))) {
      stray = true;
    }
    if (stray || !chunks.length) {
      kept.push(raw);
      continue;
    }
    let localForeman = "";
    let localPeer = "";
    let unknown = false;
    for (const chunk of chunks) {
      const title = chunk.title.trim().toLowerCase();
      if (title === "foreman") {
        localForeman = chunk.body;
      } else if (title === "peer") {
        localPeer = chunk.body;
      } else {
        unknown = true;
        break;
      }
    }
    if (unknown) {
      kept.push(raw);
      continue;
    }
    foreman = localForeman;
    peer = localPeer;
    used = true;
  }

  return {
    common: kept.join("\n\n").trim(),
    foreman,
    peer,
    used,
  };
}

export function parseHelpMarkdown(markdown: string): ParsedHelpMarkdown {
  const sections = splitSections(markdown);
  const commonSections: string[] = [];
  const actorNotes: Record<string, string> = {};
  const extraTaggedBlocks: string[] = [];
  let foreman = "";
  let peer = "";

  for (const section of sections) {
    const raw = trimBlock(section);
    if (!raw) continue;
    const tagged = parseTaggedSection(raw);
    if (!tagged) {
      commonSections.push(raw);
      continue;
    }
    if (tagged.kind === "role") {
      if (tagged.key === "role:foreman") foreman = tagged.body;
      else if (tagged.key === "role:peer") peer = tagged.body;
      else extraTaggedBlocks.push(tagged.raw);
      continue;
    }
    if (tagged.kind === "actor") {
      const actorId = tagged.key.slice("actor:".length);
      if (actorId) actorNotes[actorId] = tagged.body;
      else extraTaggedBlocks.push(tagged.raw);
      continue;
    }
    extraTaggedBlocks.push(tagged.raw);
  }

  let common = commonSections.join("\n\n").trim();
  let usedLegacyRoleNotes = false;
  if (!foreman && !peer) {
    const legacy = tryExtractLegacyRoleNotes(common);
    common = legacy.common;
    if (legacy.foreman) foreman = legacy.foreman;
    if (legacy.peer) peer = legacy.peer;
    usedLegacyRoleNotes = legacy.used;
  }

  return { common, foreman, peer, actorNotes, extraTaggedBlocks, usedLegacyRoleNotes };
}

export function buildHelpMarkdown(input: {
  common: string;
  foreman: string;
  peer: string;
  actorNotes: Record<string, string>;
  actorOrder?: string[];
  extraTaggedBlocks?: string[];
}): string {
  const parts: string[] = [];
  const common = trimBlock(input.common);
  const foreman = trimBlock(input.foreman);
  const peer = trimBlock(input.peer);
  const actorNotes = input.actorNotes || {};
  const extraTaggedBlocks = Array.isArray(input.extraTaggedBlocks) ? input.extraTaggedBlocks.map(trimBlock).filter(Boolean) : [];

  if (common) parts.push(common);
  if (foreman) parts.push(`## @role: foreman\n\n${foreman}`);
  if (peer) parts.push(`## @role: peer\n\n${peer}`);

  const seen = new Set<string>();
  const orderedActorIds: string[] = [];
  for (const actorId of Array.isArray(input.actorOrder) ? input.actorOrder : []) {
    const id = String(actorId || "").trim();
    if (!id || seen.has(id)) continue;
    seen.add(id);
    orderedActorIds.push(id);
  }
  for (const actorId of Object.keys(actorNotes).sort()) {
    if (seen.has(actorId)) continue;
    seen.add(actorId);
    orderedActorIds.push(actorId);
  }
  for (const actorId of orderedActorIds) {
    const body = trimBlock(actorNotes[actorId]);
    if (!body) continue;
    parts.push(`## @actor: ${actorId}\n\n${body}`);
  }
  parts.push(...extraTaggedBlocks);
  const out = parts.filter(Boolean).join("\n\n").trim();
  return out ? `${out}\n` : "";
}

export function updateActorHelpNote(markdown: string, actorId: string, note: string, actorOrder?: string[]): string {
  const parsed = parseHelpMarkdown(markdown);
  const nextActorNotes = { ...parsed.actorNotes };
  const aid = String(actorId || "").trim();
  if (aid) nextActorNotes[aid] = trimBlock(note);
  if (aid && !nextActorNotes[aid]) delete nextActorNotes[aid];
  return buildHelpMarkdown({
    common: parsed.common,
    foreman: parsed.foreman,
    peer: parsed.peer,
    actorNotes: nextActorNotes,
    actorOrder,
    extraTaggedBlocks: parsed.extraTaggedBlocks,
  });
}
